subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1250 @@
+		 "Module task has NULL process function pointer");
+
+	/* Execute thread */
+	while (1) {
+		data = NULL;
+
+		/* Get a new output buffer.
+		 * Since this input module generates data within itself, the module itself
+		 * will control the data flow.
+		 */
+		ret = k_mem_slab_alloc(handle->thread.data_slab, (void **)&data, K_NO_WAIT);
+		__ASSERT(ret, "No free data for module %s, ret %d", handle->name, ret);
+
+		/* Configure new audio data. */
+		audio_data.data = data;
+		audio_data.data_size = handle->thread.data_size;
+
+		/* Process the input audio data */
+		ret = handle->description->functions->data_process(
+			(struct audio_module_handle_private *)handle, NULL, &audio_data);
+		if (ret) {
+			k_mem_slab_free(handle->thread.data_slab, (void **)(&data));
+
+			LOG_ERR("Data process error in module %s, ret %d", handle->name, ret);
+			continue;
+		}
+
+		LOG_DBG("Module %s received new audio data ", handle->name);
+
+		/* Send input audio data to next module(s). */
+		send_to_connected_modules(handle, &audio_data);
+	}
+
+	CODE_UNREACHABLE;
+}
+
+/**
+ * @brief The thread that processes inputs and outputs them out of the audio system.
+ *
+ * @param handle[in/out]  The handle for this modules instance.
+ *
+ * @return 0 if successful, error value.
+ */
+static void module_thread_output(struct audio_module_handle *handle, void *p2, void *p3)
+{
+	int ret;
+
+	struct audio_module_message *msg_rx;
+	size_t size;
+
+	__ASSERT(handle != NULL, "Module task has NULL handle");
+	__ASSERT(handle->description->functions->data_process != NULL,
+		 "Module task has NULL process function pointer");
+
+	/* Execute thread. */
+	while (1) {
+		msg_rx = NULL;
+
+		LOG_DBG("Module %s waiting for audio data", handle->name);
+
+		/* Get a new input message.
+		 * Since this input message is queued outside the module, this will then control the
+		 * data flow.
+		 */
+		ret = data_fifo_pointer_last_filled_get(handle->thread.msg_rx, (void **)&msg_rx,
+							&size, K_FOREVER);
+		__ASSERT(ret, "Module %s error in getting last filled", handle->name);
+
+		LOG_DBG("Module %s new audio data received", handle->name);
+
+		/* Process the input audio data and output from the audio system. */
+		ret = handle->description->functions->data_process(
+			(struct audio_module_handle_private *)handle, &msg_rx->audio_data, NULL);
+		if (ret) {
+			if (msg_rx->response_cb != NULL) {
+				msg_rx->response_cb(
+					(struct audio_module_handle_private *)msg_rx->tx_handle,
+					&msg_rx->audio_data);
+			}
+
+			LOG_ERR("Data process error in module %s, ret %d", handle->name, ret);
+			continue;
+		}
+
+		if (msg_rx->response_cb != NULL) {
+			msg_rx->response_cb((struct audio_module_handle_private *)msg_rx->tx_handle,
+					    &msg_rx->audio_data);
+		}
+
+		data_fifo_block_free(handle->thread.msg_rx, (void **)&msg_rx);
+	}
+
+	CODE_UNREACHABLE;
+}
+
+/**
+ * @brief The thread that processes inputs and outputs the data from the module.
```

Comment:
So this is only used if data is flowing between modules?

Reply:
Well this is a module that receives data processes it and then can send to another module or back to the application.